### PR TITLE
backend/feat: added merchant/partner creation api

### PR DESCRIPTION
**Pull request by AshishSaini25**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

added merchant/partner creation api

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Merchant.hs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  TokenInfo ->
+  CreateMerchantReq ->
+  m DMerchant.MerchantAPIEntity
+createMerchant _ req = do
+  let shortId = ShortId req.shortId :: ShortId DMerchant.Merchant
+  mbExistingMerchant <- QMerchant.findByShortId shortId
+  whenJust mbExistingMerchant $ \_ -> throwError (MerchantAlreadyExist req.shortId)
+  merchant <- buildMerchant req
+  decMerchant <- decrypt merchant
+  Esq.runTransaction $
```
Comment (0utkarsh) [inline Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Merchant.hs:75]: transaction not required

File: Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Merchant.hs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  merchantUserAccountLimit <- asks (.merchantUserAccountNumber)
+  when (length merchantAssociatedAccount >= merchantUserAccountLimit) $ throwError (MerchantAccountLimitExceeded (merchant.shortId.getShortId))
+  role <- QRole.findByName req.roleName >>= fromMaybeM (RoleDoesNotExist req.roleName)
+  let personCreateReq = DPerson.CreatePersonReq {roleId = role.id, email = req.email, password = req.password, firstName = req.firstName, lastName = req.lastName, mobileNumber = req.mobileNumber, mobileCountryCode = req.mobileCountryCode}
+  personEntity <- DPerson.createPerson tokenInfo personCreateReq
+  mbMerchantAccess <- QAccess.findByPersonIdAndMerchantIdAndCity personEntity.person.id merchant.id tokenInfo.city
+  whenJust mbMerchantAccess $ \_ -> do
+    throwError $ InvalidRequest "Merchant access already assigned."
+  merchantAccess <- DPerson.buildMerchantAccess personEntity.person.id merchant.id merchant.shortId tokenInfo.city
+  Esq.runTransaction $
```
Comment (0utkarsh) [inline Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Merchant.hs:135]: same

File: Backend/app/dashboard/Lib/src/Domain/Types/Merchant.hs [hunk 3]
```diff
@@ -11,21 +12,78 @@
 
  the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 -}
+{-# LANGUAGE InstanceSigs #-}
```
Comment (0utkarsh) [inline Backend/app/dashboard/Lib/src/Domain/Types/Merchant.hs:15]: Above pragmas look unused

File: Backend/app/dashboard/Lib/src/Domain/Types/Merchant.hs [hunk 4]
```diff
@@ -11,21 +12,78 @@
...
+    HasField (hasField),
+    Maybe,
+    Monad (return),
+    Show,
+    Text,
+    ToJSON,
+    ToSchema,
+    UTCTime,
+    fst,
+    ($),
```
Comment (0utkarsh) [inline Backend/app/dashboard/Lib/src/Domain/Types/Merchant.hs:35]: Why all this separately?

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

File: Backend/app/dashboard/Lib/src/Domain/Types/Merchant.hs (hunk 3 above)
Comment (AshishSaini25) [reply Backend/app/dashboard/Lib/src/Domain/Types/Merchant.hs:15]: No , below there is a  explicit type signatures in instance declarations.

```

File: Backend/app/dashboard/Lib/src/API/Dashboard/Merchant.hs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    :> "merchant"
+    :> ( "create"
+           :> DashboardAuth 'DASHBOARD_ADMIN
+           :> ReqBody '[JSON] DMerchant.CreateMerchantReq
+           :> Post '[JSON] DMerchant.MerchantAPIEntity
+           :<|> DashboardAuth 'DASHBOARD_ADMIN
+             :> "merchantList"
+             :> Get '[JSON] [DMerchant.MerchantAPIEntity]
+           :<|> DashboardAuth 'MERCHANT_ADMIN
+             :> "createUserForMerchant"
```
Comment (hkmangla) [inline Backend/app/dashboard/Lib/src/API/Dashboard/Merchant.hs:37]: Just write "create/user".

File: Backend/app/dashboard/Lib/src/API/Dashboard/Merchant.hs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+type API =
+  "admin"
+    :> "merchant"
+    :> ( "create"
+           :> DashboardAuth 'DASHBOARD_ADMIN
+           :> ReqBody '[JSON] DMerchant.CreateMerchantReq
+           :> Post '[JSON] DMerchant.MerchantAPIEntity
+           :<|> DashboardAuth 'DASHBOARD_ADMIN
+             :> "merchantList"
```
Comment (hkmangla) [inline Backend/app/dashboard/Lib/src/API/Dashboard/Merchant.hs:34]: "list"

File: Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Merchant.hs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+import Tools.Auth
+import Tools.Auth.Merchant
+import Tools.Error
+
+data CreateMerchantReq = CreateMerchantReq
+  { shortId :: Text,
+    is2faMandatory :: Bool,
+    defaultOperatingCity :: City.City,
+    supportedOperatingCities :: [City.City],
+    serverNames :: [DTServer.ServerName],
```
Comment (hkmangla) [inline Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Merchant.hs:43]: Don't take input from users. It can be hardcoded.

**Review by hkmangla — CHANGES_REQUESTED** [review]

**Review by hkmangla — APPROVED** [review]

**Pull request merged**